From afbac29a2c97f05e831d2cc31f91cfd3a0183e5c Mon Sep 17 00:00:00 2001
From: Viorel Suman <viorel.suman@nxp.com>
Date: Mon, 12 Feb 2018 12:35:19 +0200
Subject: [PATCH 3424/5242] MLK-15071: ASoC: fsl: imx-ak4497: Fix clk for
 384KHz and 786KHz

commit  94b4ed3160646be2d511a609bebfbd1c9fecc4b0 from
https://source.codeaurora.org/external/imx/linux-imx.git

With the current multipliers SAI isn't able to derive a correct bitclk.

e.g: When playing at 786Khz with current multiplier

MCLK = 22579200, requested freq 22579200 but SAI wants:
MCLK = (DIV + 1) * 2 * freq [SAI TCR2], so an acceptable solution
is to add a 2x factor to mclk.

Signed-off-by: Viorel Suman <viorel.suman@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/codecs/ak4497.c  |    3 ++-
 sound/soc/fsl/imx-ak4497.c |    4 ++--
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/sound/soc/codecs/ak4497.c b/sound/soc/codecs/ak4497.c
index a184755..dc4d529 100644
--- a/sound/soc/codecs/ak4497.c
+++ b/sound/soc/codecs/ak4497.c
@@ -544,16 +544,17 @@ static int ak4497_hw_params(struct snd_pcm_substream *substream,
 			dfs |= AK4497_DFS_96KHZ;
 			dfs2 |= AK4497_DFS2_48KHZ;
 			break;
-
 		case 176400:
 		case 192000:
 			dfs |= AK4497_DFS_192KHZ;
 			dfs2 |= AK4497_DFS2_48KHZ;
 			break;
+		case 352800:
 		case 384000:
 			dfs |= AK4497_DFS_384KHZ;
 			dfs2 |= AK4497_DFS2_384KHZ;
 			break;
+		case 705600:
 		case 768000:
 			dfs |= AK4497_DFS_768KHZ;
 			dfs2 |= AK4497_DFS2_384KHZ;
diff --git a/sound/soc/fsl/imx-ak4497.c b/sound/soc/fsl/imx-ak4497.c
index b1bd626..71dee45 100644
--- a/sound/soc/fsl/imx-ak4497.c
+++ b/sound/soc/fsl/imx-ak4497.c
@@ -45,8 +45,8 @@ struct imx_ak4497_data {
 	{ .min = 44100,  .max = 48000,  .mul = 512  }, /* Normal */
 	{ .min = 88200,  .max = 96000,  .mul = 256  }, /* Double */
 	{ .min = 176400, .max = 192000, .mul = 128  }, /* Quad */
-	{ .min = 384000, .max = 384000, .mul = 64   }, /* Oct */
-	{ .min = 768000, .max = 768000, .mul = 32   }, /* Hex */
+	{ .min = 352800, .max = 384000, .mul = 2*64 }, /* Oct */
+	{ .min = 705600, .max = 768000, .mul = 2*32 }, /* Hex */
 };
 
 static bool imx_ak4497_is_dsd(struct snd_pcm_hw_params *params)
-- 
1.7.9.5

